### PR TITLE
#286: Removed Spray

### PR DESCRIPTION
**Pull request by akara**
Thanks for your pull request.  Please review the following guidelines.

- [X] Title includes issue id.
- [X] Description of the change added.
- [X] Commits are squashed.
- [ ] Tests added.
- [ ] Documentation added/updated.
- [X] Also please review [CONTRIBUTING.md](https://github.com/paypal/squbs/blob/master/CONTRIBUTING.md).

Refactored serialization artifacts to Akka-HTTP. Removed artifacts with open source duplication.
While this PR makes many fixes for #329, it does not complete #329. We still need to work on all Java aspects of serialization, complete test cases, and complete serialization documentation. Please review only on the merit of Spray removal and recovering existing functionality with the serializers that would otherwise be lost.

### CODE REVIEW

File: squbs-httpclient/src/test/scala/org/squbs/httpclient/dummy/DummyEndpointResolver.scala [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import org.squbs.httpclient.HttpClientException
+
+class DummyServiceEndpointResolver(dummyServiceEndpoint: String)
+                                  (implicit system: ActorSystem) extends EndpointResolver {
+
+  override def resolve(svcName: String, env: Environment): Option[Endpoint] = {
+    if (svcName == name) {
+      Some(Endpoint(dummyServiceEndpoint))
+    } else {
+      None
```
Comment (anilgursel) [inline squbs-httpclient/src/test/scala/org/squbs/httpclient/dummy/DummyEndpointResolver.scala]: I probably need to make `EndpointResolver` more Java friendly.

File: squbs-httpclient/src/test/scala/org/squbs/httpclient/dummy/DummyEndpointResolver.scala [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      case _   => throw new HttpClientException("DummyLocalhostResolver cannot support " + env + " environment")
+    }
+  }
+
+  override def name: String = "DummyLocalhostResolver"
+}
+
+object GoogleAPI {
+
+  class GoogleMapAPIEndpointResolver(implicit system: ActorSystem) extends EndpointResolver {
```
Comment (anilgursel) [inline squbs-httpclient/src/test/scala/org/squbs/httpclient/dummy/DummyEndpointResolver.scala]: I, on purpose, removed any "demo" code from the tests.  At least, that's what I remember.  AFAIR, `GoogleAPI` is used by one of those "demo"s.  I would suggest eliminating this.  We should not have "demo" code, or anything that requires to run a `main` in the test..  We should be self container and just rely on our tests.  Accordingly, if there is no test usage of `GoogleAPI`, I suggest to remove it.

File: squbs-httpclient/src/test/scala/org/squbs/httpclient/dummy/DummyService.scala [hunk 3]
```diff
@@ -19,23 +19,30 @@ package org.squbs.httpclient.dummy
 import java.util
 
 import akka.actor.ActorSystem
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.Http.ServerBinding
+import akka.http.scaladsl.marshalling.{Marshaller, ToEntityMarshaller}
+import akka.http.scaladsl.model.headers.RawHeader
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.server.Route
```
Comment (anilgursel) [inline squbs-httpclient/src/test/scala/org/squbs/httpclient/dummy/DummyService.scala]: Probably need to run "Optimize imports" to fix the codacy issue.

File: squbs-httpclient/src/test/scala/org/squbs/httpclient/dummy/DummyService.scala [hunk 4]
```diff
@@ -19,23 +19,30 @@ package org.squbs.httpclient.dummy
...
+import akka.http.scaladsl.server.Directives
+import akka.http.scaladsl.unmarshalling.FromEntityUnmarshaller
+import akka.stream.ActorMaterializer
+import com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility
+import com.fasterxml.jackson.annotation.PropertyAccessor
 import com.fasterxml.jackson.databind.ObjectMapper
 import com.fasterxml.jackson.module.scala.DefaultScalaModule
+import org.json4s.{DefaultFormats, jackson}
```
Comment (anilgursel) [inline squbs-httpclient/src/test/scala/org/squbs/httpclient/dummy/DummyService.scala]: ditto

File: squbs-httpclient/src/test/scala/org/squbs/httpclient/dummy/DummyService.scala [hunk 5]
```diff
@@ -86,20 +93,14 @@ object EmployeeBeanSerializer extends CustomSerializer[EmployeeBean](format => (
...
 
-*/
-object DummyService {
-
-  val (dummyServiceIpAddress, dummyServicePort) = temporaryServerHostnameAndPort()
-
-  val dummyServiceEndpoint = Uri(s"http://$dummyServiceIpAddress:$dummyServicePort")
-}
-/*
 object DummyServiceMain extends App with DummyService {
```
Comment (anilgursel) [inline squbs-httpclient/src/test/scala/org/squbs/httpclient/dummy/DummyService.scala]: We should eliminate these..  I do not think we should have any `main` method.  They easily get out of sync.

File: squbs-httpclient/src/test/scala/org/squbs/httpclient/json/Json4sJacksonSpec.scala [hunk 6]
```diff
@@ -16,71 +16,73 @@
...
+
+  import de.heikoseeberger.akkahttpjson4s.Json4sSupport._
+
+  implicit val system = ActorSystem("Json4sJacksonSpec")
+  implicit val mat = ActorMaterializer()
+  implicit val serialization = jackson.Serialization
 
   "NotTypeHints Example (case class)" should "have correct behaviour of read/write" in {
-    import Json4sJacksonNoTypeHintsProtocol._
+    implicit val formats = DefaultFormats.withHints(NoTypeHints)
```
Comment (anilgursel) [inline squbs-httpclient/src/test/scala/org/squbs/httpclient/json/Json4sJacksonSpec.scala]: `DefaultFormats.withHints(NoTypeHints)` looks confusing to me.  Should we rename the API as `withTypeHints` and `withoutTypeHints` or something.  Just "hints" is not clear, IMO.

Comment (akara) [inline squbs-httpclient/src/test/scala/org/squbs/httpclient/json/Json4sJacksonSpec.scala]: This is not our API. It is from Json4s. But I can just drop the `.withHints(...)` in this case. It just explicitly mentions no type hints should be rendered.

File: squbs-httpclient/src/test/scala/org/squbs/httpclient/dummy/DummyEndpointResolver.scala (hunk 1 above)
Comment (akara) [inline squbs-httpclient/src/test/scala/org/squbs/httpclient/dummy/DummyEndpointResolver.scala]: Yup.

File: squbs-httpclient/src/main/scala/org/squbs/httpclient/json/XLangJsonSupport.scala [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+                                                    formats: Formats = DefaultFormats): FromEntityUnmarshaller[R] =
+      if (isJavaClass(clazz)) {
+        implicit val classTag = ClassTag[R](clazz)
+        JacksonMapperSupport.jacksonUnmarshaller[R]
+      } else {
+        implicit val manifest = ManifestFactory.classType[R](clazz)
+        Json4sSupport.json4sUnmarshaller[R]
+      }
+
+//    private implicit def unmarshallerToFromResponseUnmarshaller[T](ummarshaller: Unmarshaller[T]): FromResponseUnmarshaller[T] = {
```
Comment (akara) [inline squbs-httpclient/src/main/scala/org/squbs/httpclient/json/XLangJsonSupport.scala]: I want to keep all of these until we have the Java test cases all done.

File: squbs-httpclient/src/main/scala/org/squbs/httpclient/json/XLangJsonSupport.scala [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        JacksonMapperSupport.jacksonMarshaller[T]
+      } else {
+        Json4sSupport.json4sMarshaller[T]
+      }
+    }
+
+    implicit def typeTagToUnmarshaller[R: TypeTag](implicit
+                                                    serialization: Serialization = jackson.Serialization,
+                                                    formats: Formats = DefaultFormats): FromEntityUnmarshaller[R] = {
+      if (typeOf[R].typeSymbol.isJava) {
```
Comment (akara) [inline squbs-httpclient/src/main/scala/org/squbs/httpclient/json/XLangJsonSupport.scala]: I have to check. AFAIK, it should be only once when the route is built.

File: squbs-httpclient/src/main/scala/org/squbs/httpclient/json/XLangJsonSupport.scala [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    //register a specific mapper
+    def registerMapper(clazz: Class[_], mapper: ObjectMapper): Unit = {
+      synchronized {
+        mappers = mappers + ((clazz.getName, mapper))
+      }
+    }
+
+    //default mapper relies on getters
+    val defaultMapper = new ObjectMapper()
+    //.setVisibility(PropertyAccessor.FIELD, Visibility.ANY)
```
Comment (akara) [inline squbs-httpclient/src/main/scala/org/squbs/httpclient/json/XLangJsonSupport.scala]: You mean the comments? This is old code. We had to remove as they cause issues. I can remove them. But before doing so, I'd like to get all the usage Java and Scala in place.

File: squbs-httpclient/build.sbt [hunk 10]
```diff
@@ -7,23 +7,26 @@ javaOptions in Test += "-Xmx512m"
 libraryDependencies ++= Seq(
   "com.typesafe.akka"         %% "akka-actor"                   % akkaV,
   "com.typesafe.akka"         %% "akka-slf4j"                   % akkaV,
-  "com.typesafe.akka"         %% "akka-http-core"               % akkaHttpV ,
+  "com.typesafe.akka"         %% "akka-http"                    % akkaHttpV ,
```
Comment (akara) [inline squbs-httpclient/build.sbt]: Hmm, good point. Don't recall this change. But I tried changing it back and all still compiles. So we should be good setting this back.

**Pull request merged**